src/browser/providers/tidal.js

```diff
@@ -0,0 +1,267 @@
+            this.controls.play &&
+            this.controls.play.click()) ||
+            super.play();
+    }
+
+    pause() {
+        (this.activePlayer.isPlaying() &&
+            this.controls.play &&
+            this.controls.play.click()) ||
+            super.pause();
+    }
+
+    playPause() {
+        (this.controls.play && this.controls.play.click()) || super.playPause();
+    }
+
+    setShuffle(isShuffle) {
+        if (this.controls.shuffle) {
+            if (
+                (!this.isShuffle() && isShuffle) ||
+                (this.isShuffle() && !isShuffle)
+            ) {
+                this.controls.shuffle.click();
+            }
+        }
+    }
+
+    isShuffle() {
+        if (this.controls.shuffle) {
+            const classes = Array.from(this.controls.shuffle.classList);
+            return classes.some(c => c.startsWith('active'));
+        } else {
+            return super.isShuffle();
+        }
+    }
+
+    getLoopStatus() {
+        if (this.controls.repeat) {
+            const classes = Array.from(this.controls.repeat.classList);
+            if (classes.some(c => c.startsWith('once'))) {
+                return LoopStatus.TRACK;
+            } else if (classes.some(c => c.startsWith('all'))) {
+                return LoopStatus.PLAYLIST;
+            } else {
+                return LoopStatus.NONE;
+            }
+        } else {
+            return super.getLoopStatus();
+        }
+    }
+
+    setLoopStatus(status) {
+        if (this.controls.repeat) this.controls.repeat.click();
+        else super.setLoopStatus(status);
+    }
+
+    setVolume(volume) {
+        super.setVolume(volume);
+        if (this.controls.volumeProgress) {
+            this.controls.volumeProgress.value = volume * 100;
+            this.controls.volumeProgress.style =
+                '--val:' + volume * 100 + '; --max:100; --min:0;';
+        }
+    }
+
+    getVolume() {
+        if (this.controls.volumeProgress) {
+            return parseInt(this.controls.volumeProgress.value, 10) / 100;
+        }
+        return super.getVolume();
+    }
+
+    setRate() {
+        // disabled
+    }
+}
+
+Playback = TidalPlayback;
+
+window.addEventListener('mpris2-setup', function() {
+    /*
+     * Since tidal is loaded dynamically, we need to wait until the playback controls are added
+     * to the DOM before we can get the elements.
+     *
+     * We create an observer to monitor the added nodes, and set the playback controls when the
+     * player is loaded.
+     *
+     */
+    const playerObserver = new MutationObserver(mutations => {
+        let loaded = false;
+        mutations.forEach(mutation => {
+            mutation.addedNodes.forEach(n => {
+                if (n.className && n.className.startsWith('footerPlayer')) {
+                    loaded = true;
+                    playerObserver.disconnect()
+                }
+            });
+        });
+
+        if (loaded) {
```

Comment:
Since you are disconnecting the observer, couldn't we move this if's block to where we set the `loaded` to `true` and get rid of this flag? It could all be inside a initControls function or something.

Reply:
Yeah I guess that's a bit cleaner.